linkerd/app/inbound/src/lib.rs

```diff
@@ -115,12 +116,20 @@ impl<A: OrigDstAddr> Config<A> {
                     move |_| Ok(backoff.stream())
                 }))
                 .into_new_service()
-                .push_on_response(
-                    svc::layers()
-                        .push_lock()
-                        .push(metrics.stack.layer(stack_labels("endpoint"))),
+                .cache(
+                    svc::layers().push_on_response(
+                        svc::layers()
+                            // If the service has been ready & unused for `cache_max_idle_age`,
+                            // fail it.
+                            .push_idle_timeout(cache_max_idle_age)
+                            // If the service has been unavailable for an extend time, eagerly
+                            // fail requests.
+                            .push_failfast(Duration::from_secs(10))
+                            // Shares the service, ensuring discovery errors are propagated.
+                            .push_spawn_buffer(10)
```

Comment:
These values should at least be constified, and probably exposed via configuration